src/locator.rs

```diff
@@ -65,55 +63,84 @@ macro_rules! locator {
     };
 }
 
-/// Core, and most services that interact with Core,
-/// refer to open source packages via the `Locator` type.
+/// The regular expression used to parse the locator.
 ///
-/// This type is nearly universally rendered to a string
-/// before being serialized to the database or sent over the network.
+/// ```
+/// # use locator::locator_regex;
 ///
-/// This type represents a _validly-constructed_ `Locator`, but does not
-/// validate whether a `Locator` is actually valid. This means that a
-/// given `Locator` is guaranteed to be correctly formatted data,
-/// but that the actual repository or revision to which the `Locator`
-/// refers is _not_ guaranteed to exist or be accessible.
-/// Currently the canonical method for validating whether a given `Locator` is
-/// accessible is to run it through the Core fetcher system.
+/// // Get the raw string used for the expression.
+/// let expression = locator_regex!();
+///
+/// // Parse the regular expression.
+/// // The expression is compiled once per callsite.
+/// let parsed = locator_regex!(parse => "git+github.com/fossas/locator-rs$v2.2.0");
+/// ```
+#[macro_export]
+#[doc(hidden)]
+macro_rules! locator_regex {
+    () => {
+        r"^(?:([a-z-]+)\+|)([^$]+)(?:\$|)(.+|)$"
```

Comment:
Why not make this a `const` instead of repeating it?

Reply:
`lazy_regex::regex_captures` requires a string literal. Colocating like this so that it is obvious if they're out of sync was the closest way to get the "constant like" behavior.